packages/docusaurus-theme-common/src/contexts/colorMode.tsx

```diff
@@ -36,33 +38,71 @@ const Context = React.createContext<ContextValue | undefined>(undefined);
 const ColorModeStorageKey = 'theme';
 const ColorModeStorage = createStorageSlot(ColorModeStorageKey);
 
+const ColorModeChoices = {
+  auto: 'auto',
+  light: 'light',
+  dark: 'dark',
+} as const;
+
+export type ColorModeChoice =
+  typeof ColorModeChoices[keyof typeof ColorModeChoices];
+
 const ColorModes = {
   light: 'light',
   dark: 'dark',
 } as const;
 
 export type ColorMode = typeof ColorModes[keyof typeof ColorModes];
 
-// Ensure to always return a valid colorMode even if input is invalid
-const coerceToColorMode = (colorMode?: string | null): ColorMode =>
-  colorMode === ColorModes.dark ? ColorModes.dark : ColorModes.light;
+// Ensure to always return a valid colorModeChoice even if input is invalid
+const coerceToColorMode = (
+  colorModeChoice?: string | null,
+): ColorModeChoice => {
+  switch (colorModeChoice) {
+    case ColorModeChoices.light:
+      return ColorModeChoices.light;
+    case ColorModeChoices.dark:
+      return ColorModeChoices.dark;
+    case ColorModeChoices.auto:
+    default:
+      return ColorModeChoices.auto;
+  }
+};
 
-const getInitialColorMode = (defaultMode: ColorMode | undefined): ColorMode =>
-  ExecutionEnvironment.canUseDOM
-    ? coerceToColorMode(document.documentElement.getAttribute('data-theme'))
-    : coerceToColorMode(defaultMode);
+const getInitialColorMode = (
+  defaultMode: ColorModeChoice | undefined,
+): ColorModeChoice =>
+  coerceToColorMode(
+    ExecutionEnvironment.canUseDOM
+      ? document.documentElement.getAttribute('data-theme-choice') ??
+          document.documentElement.getAttribute('data-theme')
+      : defaultMode,
+  );
 
-const storeColorMode = (newColorMode: ColorMode) => {
+const storeColorMode = (newColorMode: ColorModeChoice) => {
   ColorModeStorage.set(coerceToColorMode(newColorMode));
 };
 
 function useContextValue(): ContextValue {
   const {
-    colorMode: {defaultMode, disableSwitch, respectPrefersColorScheme},
+    colorMode: {defaultMode, disableSwitch},
   } = useThemeConfig();
-  const [colorMode, setColorModeState] = useState(
+  const [colorModeChoice, setColorModeState] = useState(
     getInitialColorMode(defaultMode),
   );
+  const colorMode = useMemo(() => {
+    switch (colorModeChoice) {
+      case ColorModeChoices.light:
+        return ColorModes.light;
+      case ColorModeChoices.dark:
+        return ColorModes.dark;
+      case ColorModeChoices.auto:
+      default:
+        return window.matchMedia('(prefers-color-scheme: dark)').matches
```

Comment:
it's not possible to access window when rendering React components (and useMemo is executed when rendering).

When doing SSR/SSG, window does not exist and it's impossible to query for prefers color scheme because we are not inside a browser.

The error maybe doesn't explain this well, but with experience we figure out the problem quite fast:

```
6:29:39 AM: [ERROR] Docusaurus server-side rendering could not render static page with path /docs/2.1.0/creating-pages/.
6:29:39 AM: [INFO] It looks like you are using code that should run on the client-side only.
6:29:39 AM: To get around it, try using `<BrowserOnly>` (https://docusaurus.io/docs/docusaurus-core/#browseronly) or `ExecutionEnvironment` (https://docusaurus.io/docs/docusaurus-core/#executionenvironment).
6:29:39 AM: It might also require to wrap your client code in `useEffect` hook and/or import a third-party library dynamically (if any).
6:30:15 AM: ReferenceError: window is not defined
```

Reply:
oooh that makes a lot of sense yeah

so what's the plan here, since we cannot know in advance what theme to render with from the server-side? do i just assume light when `window` isn't available or is there something smarter to do?

Reply:
@nasso We use `defaultTheme`.

Reply:
the thing is, `defaultMode` now includes `auto`, but here we need to figure out the effective theme (`light` or `dark`)

Reply:
It's not possible to figure out the effective theme during the SSR render phase and unfortunately will never be possible. There's no theme when running in Node.js. That's why we use inlined JS and the data-theme attribute to set the effective theme as soon as we can to avoid a "theme flash". Doing this in render/useMemo is both too late and not possible. Doing this in useEffect is possible but also too late to avoid the flash.

I think there's already something we do in existing code that is even not compatible with React 18. The server render must **match exactly** the first client render (hydration). And we can't get the effective theme on the server, so both the server/first-client render should use the exact same value (the default theme, not the effective one).

I mean: this existing code is likely wrong because it can lead to different values for the server render and the first client render:

```tsx
const getInitialColorMode = (
  defaultMode: ColorModeChoice | undefined,
): ColorModeChoice =>
  coerceToColorMode(
    ExecutionEnvironment.canUseDOM
      ? document.documentElement.getAttribute('data-theme-choice') ??
          document.documentElement.getAttribute('data-theme')
      : defaultMode,
  );
```

Reply:
i think one way we could get rid of the FOUC altogether would be to handle `auto` as a possible value for the `data-theme` attribute, and use `@media` queries in CSS to set the theme instead of dynamically querying the preference using JS. CSS media queries would only apply on `[data-theme=auto]`. this would probably require modifying a lot of CSS though.

Reply:
That's a possibility, but I think it will duplicate CSS, and also be annoying for users that already have custom dark/light mode elements.

We/they would now have to use more complex selectors everywhere:
- `@media (prefers-color-scheme: dark) [data-theme=auto] .subselector, [data-theme=dark] .subselector { ... }`
- `@media (prefers-color-scheme: light) [data-theme=auto] .subselector, [data-theme=light] .subselector { ... }`

So I don't think it would be a great move, our current theming CSS DX is better:
- `[data-theme=dark] .subselector { ... }`
- `[data-theme=light] .subselector { ... }`

Reply:
that is definitely much better.

> And we can't get the effective theme on the server, so both the server/first-client render should use the exact same value (the default theme, not the effective one).

does that mean there would be no way to set the default theme to the system preference? because i don't see what the server would render when the default is "auto"